src/coll/mod.rs

```diff
@@ -381,12 +386,17 @@ where
         &self,
         filter: Document,
         options: impl Into<Option<FindOneAndDeleteOptions>>,
-    ) -> Result<Option<Document>> {
+    ) -> Result<Option<T>> {
         let mut options = options.into();
         resolve_options!(self, options, [write_concern]);
 
         let op = FindAndModify::with_delete(self.namespace(), filter, options);
-        self.client().execute_operation(op).await
+        let result = self.client().execute_operation(op).await;
+
+        result.and_then(|opt| {
```

Comment:
Rather than repeating this in each of the `find_one_and_x` method bodies, it seems like we ought to handle this in `FindAndModify::handle_response`. We'd have to make `FindAndModify` generic as part of that.